[UnifiedPDF] Embedded PDFs are blurry when zooming the containing page

aprotyas opened the pull request:
#### 18970a1eb04af74556cf9413f2363550ac2aee61
<pre>
[UnifiedPDF] Embedded PDFs are blurry when zooming the containing page
<a href="https://bugs.webkit.org/show_bug.cgi?id=274565">https://bugs.webkit.org/show_bug.cgi?id=274565</a>
<a href="https://rdar.apple.com/128290750">rdar://128290750</a>

Reviewed by Simon Fraser.

In 275712@main, we unconditionally allow the plugin&apos;s root layer to
apply page scale. This allows any embed plugin&apos;s internal scale change
to be reflected in the tiled backing.

However, that change breaks the assumption in GraphicsLayer that
there&apos;s a single layer applying page scale. This assumption meant that
when recursively commiting changes in a layer tree, GraphicsLayer only
selects (and propagates) scale factor from the furthest descendant layer
that applies page scale. We see this aligning with the behavior that
regardless of root page scaling, all painted tiles only reflect the embed
plugin&apos;s scale.

To fix this, we move away from our previous assumption of a single layer
applying page scale. Instead, as we traverse the layer tree, we
maintain a combined page scale representing the product of the scale
factor of each layer that applies page scale. This would not only account
for differences between root page scale and embed plugin scale, but also
any intermediate CSS transform scale.

* LayoutTests/compositing/plugins/pdf/pdf-in-embed-with-root-page-zoom-expected.txt: Added.
* LayoutTests/compositing/plugins/pdf/pdf-in-embed-with-root-page-zoom.html: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/resources/image.pdf: Added.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::computePositionRelativeToBase const):
(WebCore::GraphicsLayerCA::recursiveCommitChanges):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::setPageScaleFactor):

Canonical link: <a href="https://commits.webkit.org/279314@main">https://commits.webkit.org/279314@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7025196183baccf4929a0d17535c2b0cc0eb7002

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53132 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32469 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5619 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56411 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3855 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55437 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39356 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3584 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/43075 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2494 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55230 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/30231 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45861 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24205 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3189 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2014 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49080 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3347 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58007 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28273 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3294 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50477 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29493 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46087 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/49783 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/11587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30412 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7807 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29247 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1368,10 +1368,6 @@ static String mutationObserverNotificationString()
         return;
     }
 
-    RefPtr page = this->page();
```
We don't use `page` anywhere in `setPageScaleFactor`

smfr reviewed (approved):
I think you could/should make a layout test that dumps layer trees.

aprotyas committed:
[UnifiedPDF] Embedded PDFs are blurry when zooming the containing page
https://bugs.webkit.org/show_bug.cgi?id=274565
rdar://128290750

Reviewed by Simon Fraser.

In 275712@main, we unconditionally allow the plugin's root layer to
apply page scale. This allows any embed plugin's internal scale change
to be reflected in the tiled backing.

However, that change breaks the assumption in GraphicsLayer that
there's a single layer applying page scale. This assumption meant that
when recursively commiting changes in a layer tree, GraphicsLayer only
selects (and propagates) scale factor from the furthest descendant layer
that applies page scale. We see this aligning with the behavior that
regardless of root page scaling, all painted tiles only reflect the embed
plugin's scale.

To fix this, we move away from our previous assumption of a single layer
applying page scale. Instead, as we traverse the layer tree, we
maintain a combined page scale representing the product of the scale
factor of each layer that applies page scale. This would not only account
for differences between root page scale and embed plugin scale, but also
any intermediate CSS transform scale.

* LayoutTests/compositing/plugins/pdf/pdf-in-embed-with-root-page-zoom-expected.txt: Added.
* LayoutTests/compositing/plugins/pdf/pdf-in-embed-with-root-page-zoom.html: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/resources/image.pdf: Added.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::computePositionRelativeToBase const):
(WebCore::GraphicsLayerCA::recursiveCommitChanges):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::setPageScaleFactor):

Canonical link: https://commits.webkit.org/279314@main
